### DISCUSSION TITLE
MAVLINK stream via WIFI on ELRS backpack

### DISCUSSION THREAD

**Discussion by Christoph-Hofmann** [ExpressLRS/ExpressLRS#1352] [Ideas]
It would be nice to have the possibilty to see the telemetry (gps;speed;hight) on a laptop/tablet.

A subset of Mavlink should be enough for this feature.

kind regards 
Christoph

**Comment by sweebee** [comment]
Looks similar to this one #636. Although a bit "broader". BLE is more reliable and faster than wifi.

**Reply by dz0ny** [reply]
https://github.com/ExpressLRS/ExpressLRS/compare/master...dz0ny:ExpressLRS:add/ble-telemetry?expand=1 :) Almost working

**Comment by Nordle420** [comment]
yesplease

**Comment by CapnBry** [comment]
I'm also tagging #390 in this because all three Issues are requests for spitting out telemetry via Bluetooth / Wifi over various protocols.

**Reply by charles-shaju** [reply]
i am using radio master tx12 rc remote with rp1 elrs receiver, i am getting telemetry data in my rc transmitter now i want to get live telemetry to my gcs . i am using inav configurator in my gcs how to get get it . can anyone explain

**Comment by 0crap** [comment]
noplease, stick to the core functionality that makes ExpressLRS so good and forget all this telemetry crap.

**Comment by sweebee** [comment]
@0crap thats your opinion, a lot of people will disagree. The backpack is made for stuff like this. Good thing this is open source ;)

**Comment by 0crap** [comment]
> @0crap thats your opinion, a lot of people will disagree. The backpack is made for stuff like this. Good thing this is open source ;)

It sure is, nothing wrong with giving an opinion is it....

**Comment by wucke13** [comment]
Maybe it would make sense to streamline the three issues related to this feature (https://github.com/ExpressLRS/ExpressLRS/issues/390 , https://github.com/ExpressLRS/ExpressLRS/issues/636 & https://github.com/ExpressLRS/ExpressLRS/issues/672). 

# What do we need

1. A transparent, bidirectional data link, able to transmit arbitrary bytes
2. The link should be robust, e.g. a loss of a single frame should not result in a loss of transmitted information (otherwise using MAVLink for example will be close to impossible)
2. The link should be exposed as two pins on the RX side (in the for of a serial port)
3. On the TX side, the link should be exposed either as serialport (two pins, makes a HW mod for the latter easy) or via Bluetooth/WiFi/Whatever
4. The communication link needs to be fully optional. If a user does not want to use it, it should bear no extra cost during transmission than completely without it.

Completely orthogonal to that is how this is used.

# Use cases

+ MAVLink connection for QGroundControl
+ MSP connection for wireless configuration of BF/INAV
+ Literally any other application which requires a transparent, byte oriented data link

# Also interesting 

would be if this is instead of or parallel to the normal telemetry (SmartPort, ...) if that is ever supported by ELRS.
Then of course also there could be additional apps attached to ELRS which somehow process the data. This however should be out of scope for ELRS, IMO. Keeping the complexity low and only making ELRS responsible for the data link, not any data processing already might be considered as bloat by some.


Please feel free to comment anything which I missed, I'm willing to collect anything back in this comment. I'd rather prefer having a new fresh issue for this, but I do not see the point in opening a fourth issue asking for basically the same functionality.

**Reply by timtuxworth** [reply]
Antenna tracker

**Comment by 0crap** [comment]
> 
> Please feel free to comment anything which I missed, I'm willing to collect anything back in this comment. I'd rather prefer having a new fresh issue for this, but I do not see the point in opening a fourth issue asking for basically the same functionality.

Sure, what you missed is that the majority don't need this. (The ones you also don't hear on these topics.)
So please add to the list that this feature should not imply a performance hit for everybody. (Sensitivity, packet rate, ....)
That would disappoint a lot of users.

**Comment by wucke13** [comment]
@0crap I think by now everybody who read the comments above will know that you're not a fan of this feature. I do not think there is any value in it if you reiterate the same point.

However, I totally see your point regarding the performance hit for everybody. It's my understanding that ELRS is very configurable, and this should be no exception.

**Comment by 0crap** [comment]
@wucke13 It has nothing to do with me, it's about consequences that new features might have.
I just replied on topic to your question if you missed anything. So don't give me that.

The second part of your reply is on topic, thank you.

**Comment by jonas-koeritz** [comment]
Without extra hardware on the TX side of things getting yaapu's telemetry script to work with Ardupilot on the copter through ExpressLRS would be a game changer for me. I'm currently using TBS Tracer for two of my quads because of this. I just need the Log messages to see what the thing is doing, no need (for now) to get a full Mavlink stream through ELRS, but that would be perfect of course.
To not interfere with other people's use cases this should be hidden behind a config toggle, I would happily give up some control rate for a nice, high bandwidth telemetry link.

**Comment by Moksh-2000** [comment]
> > Please feel free to comment anything which I missed, I'm willing to collect anything back in this comment. I'd rather prefer having a new fresh issue for this, but I do not see the point in opening a fourth issue asking for basically the same functionality.
> 
> Sure, what you missed is that the majority don't need this. (The ones you also don't hear on these topics.)
> So please add to the list that this feature should not imply a performance hit for everybody. (Sensitivity, packet rate, ....)
> That would disappoint a lot of users.

Hey sussy you don't know meaning of telemetry data on autopilot any RC models , it's only way to know what I'm doing and what my copter doing on air with windding with air so stop thinking this feature is not useful , @wucke13 you are absolutely right avoid this don't knowless person , and keep going on in also going to pull request but here I'm

**Comment by wucke13** [comment]
@Moksh-2000 While we seem to stay on the same side of the matter, we do not stand on the same site regarding the communication. I think its clear that the feature would be valuable for many, and harming for many others (hence it needs to be optional). I mean, there is a reason why ELRS has higher range on lower latency than FRSky Archer stuff despite both being based on the very same HW (Semtech Lora Chips).

**Comment by sweebee** [comment]
I’m sure this feature won’t harm the performance/latency in any way. This feature needs the backpack to send this data via WiFi or BLE. The backpack is currently not used at all. The main processor handles all the radio data. 

But you need a tx with a backpack ofc.

**Comment by jonas-koeritz** [comment]
I'd be happy to trade some latency for a reliable data link any time. Of course making this optional (via a build flag?) would be of help to keep the latency low for the racers. My 12" will never need a super low latency link but it would benefit greatly from Mavlink connectivity.

**Comment by wucke13** [comment]
> I’m sure this feature won’t harm the performance/latency in any way

I'm not a signals expert, but to me this looks pretty simple: If you transmit additional data (the telemetry), then you transmit more symbols (as in bits) per frame containing new controls. More symbols per frame -> requires higher bandwidth -> results in lower range if transmission power and update rate remain equal. So IMHO, the very opposite of your statement holds true: enabling (and then using this feature) would harm the performance. This has nothing to do with the processing power, but the RF link.

> I'd be happy to trade some latency for a reliable data link any time

Perfectly valid. I too. But we must recognize that this is not the only view, ELRS is special precisely as it allows low latency without sacrificing range/reliability.

No one doubts that the feature __is useful for some__, it's just __also conflicting__ with the __performance__ metrics which are __important for some others__.

**Comment by sweebee** [comment]
> > I’m sure this feature won’t harm the performance/latency in any way
> 
> I'm not a signals expert, but to me this looks pretty simple: If you transmit additional data (the telemetry), then you transmit more symbols (as in bits) per frame containing new controls. More symbols per frame -> requires higher bandwidth -> results in lower range if transmission power and update rate remain equal. So IMHO, the very opposite of your statement holds true: enabling (and then using this feature) would harm the performance. This has nothing to do with the processing power, but the RF link.
> 
> > I'd be happy to trade some latency for a reliable data link any time
> 
> Perfectly valid. I too. But we must recognize that this is not the only view, ELRS is special precisely as it allows low latency without sacrificing range/reliability.
> 
> No one doubts that the feature **is useful for some**, it's just **also conflicting** with the **performance** metrics which are **important for some others**.

Yes you’re right about additional data. But I personally don’t need mavlink, only telemetry data to my phone. For that there is no additional data needed.

**Comment by Moksh-2000** [comment]
> I'd be happy to trade some latency for a reliable data link any time. Of course making this optional (via a build flag?) would be of help to keep the latency low for the racers. My 12" will never need a super low latency link but it would benefit greatly from Mavlink connectivity.

I'm still wondering why all people need low latency because video have 20 to 50ms so why radio link below and there for nothing you can change

**Comment by jonas-koeritz** [comment]
How about just supporting CSRF-Passthrough telemetry (packet types 0x7F and 0x80) for now? This would fit into the existing telemetry bandwidth (I assume) and shouldn't have any impact on packet rates or any other ELRS qualities. Ardupilot is supporting it (in latest versions) and the yaapu script will be fine handling it.
I'm not very familiar with this codebase but I think ELRS currently just throws away packets it doesn't understand. Passthrough would be just handing them over to the TX transparently.
Maybe @yaapu could have a look at this too? I'd love to run a fully open source software and hardware stack on my copters.

**Comment by 0crap** [comment]
> I'm still wondering why all people need low latency because video have 20 to 50ms so why radio link below and there for nothing you can change

The fact that you don't understand it doesn't mean it's not a good thing.
Low and consistent latency is one of the **key features** of ExpressLRS.
We like to keep it that way.
Otherwise grab a ppm receiver and let us know your results.

Don't forget, ExpressLRS is not designed to be a replacement for TBS crossfire and the like.

**Comment by sweebee** [comment]
> > I'd be happy to trade some latency for a reliable data link any time. Of course making this optional (via a build flag?) would be of help to keep the latency low for the racers. My 12" will never need a super low latency link but it would benefit greatly from Mavlink connectivity.
> 
> I'm still wondering why all people need low latency because video have 20 to 50ms so why radio link below and there for nothing you can change

The lower the latency the better, because the latency is added. You have to account for the whole latency loop.

**Comment by markandkymward** [comment]
I too am very interested in this feature and I'm waiting to confirm it will exist before moving over.  Any consensus?

**Comment by padcom** [comment]
@0crap You say ELRS is not designed to be a replacement for Crossfire... a) how do you know that? b) does it mean it can't and won't do it? c) what is your role in the decision-making process? 

Kinda depending on where you are in all this, it should either be a compilation-time setting (at least on the RX) or a runtime setting for both RX/TX (much like wide switches recently in 2.0)

Also, there are different link capabilities that are interesting for different types of flying. For long range cinematic stuff on a self-leveling drone, you're mostly interested in having communication between your drone and the ground station. In the case of racers, freestyle and other flippidy-floppy around the tree, low latency is very important. I don't see a reason why ELRS wouldn't have different modes of operation, even one with more than 12 channels and/or transparent telemetry passthrough.

**Comment by fabi9163** [comment]
If the backpack would only outputs the telemetry from the main-esp via Wifi there wouldn't be additional delay, right?
You could connect with the smartphone and use apps like "telemetry viewer" to search for lost models or view the flightpath

**Comment by maxyB19** [comment]
This would be so great. Has someone done this yet?

**Comment by MohammadAdib** [comment]
Hi folks. I’m using betafpv elrs and thinking about switching to tbs simply because it allows me to get mavlink telemetry via wifi or bluetooth. Is there absolutely no way to get mavlink working over wifi/bl using elrs?

**Comment by ela2342** [comment]
I'd also be interested in this feature.

**Comment by janrunejohansen** [comment]
I'm also interested in this feature. I recently decided to jump on the ELRS train because of WIDE channels finally arrived and last thing that's missing for my part is the someway to pass the MAVLink data to my Surface Book 2.

**Comment by ShikOfTheRa** [comment]
+1 equivalent functionality to "Mavlink EMU on TBS"  would be awesome.

**Comment by wakesolo** [comment]
ELRS+Mavlink+BLE=OMG YES!

**Comment by commanderguy3001** [comment]
Mavlink has been stated by the developers to not get implemented multiple times by now, for numerous reasons.
Those include, as far as I remember, mostly that elrs is made to be as lightweight as possible, but mavlink itself needs a ton of bandwidth which would also drastically decrease range and responsiveness.
Also mavlink could only be done on 2.4ghz systems if I remember correctly, as 900mhz doesn't have enough bandwidth in general (I might be wrong here though).
One more thing is that it would require a lot of code rewriting and a lot more code complexity.

**Reply by commanderguy3001** [reply]
it therefore would kind of oppose the main selling points of elrs.
In general I would say: if you want mavlink, elrs isn't the right thing for you to use, take tbs or whatever in that case (though I have heard that mavlink on crossfire is a pain to get working too, but I personally have no reason to care)

**Reply by ela2342** [reply]
The main selling point of ELRS for me is that it is an open source long range solution. The ultra low latency is very nice for some of my applications (like FPV quads), but something I don't need for some other applications (long range fixed wings).

So, yes, I do understand that MAVLink would come at a price. I'm fine with that. I'd even be willing to work with a pure MAVLink only (or generic telemetry passthrough mode), as joystick over MAVLink exists.

The bandwith MAVLink requires depends on the messages one needs and their frequency, and is actually very nicely tunable. In the best of all worlds, I would be able to choose the underlying lora mode, and tune range vs. telemetry rates according to my specific needs.

I do appreciate the additional complexity argument. Every feature comes at a price. It shouldn't be too complex though.

Being able to use unified hardware would be a big win for me.

**Reply by commanderguy3001** [reply]
there are three big points to elrs:
- open source
- cheap (this is one of the top selling points, together with low latency and/or long range, depending on an individual users use-case)
- long range
- low latency

the biggest price you would pay with mavlink is a massive reduction in range, as it basically requires a full uart connection to be sent OTA, which is a lot more data to transmit compared to standard ELRS (which I think is 6 bytes per packet, so 3000 bytes a second on 500hz which is only about 24000 baud half duplex).

And while joystick over MAVlink is a thing, it is not great, and if you use it, then there really is no point to elrs anymore (it's higher latency, shorter range, only thing is it would still be cheap and open source), but in that case it would be better to start a project that only transmits uart over the air, and is possible to run on elrs hardware if you want (which would make sense to do).

The added complexity would for example come from the fact that you would have to swap between transmitting and receiving data all the time, at very specific intervals and pretty fast (which would also double the bandwidth needed now that I think about it, as RF is only half duplex, while uart is full duplex by default, but that is a problem anyway).

I don't quite understand what you mean with "unified hardware" in that regards though?

**Reply by iottrends** [reply]
I am thinking of porting the Sik Radio Telemetry on ExLRS Rx and Tx modules as a replacement for 900Mhz SiK Telemetry Radios.
My Aim is to do Auto-pilot using Ardupilot while Replacing the 900Mhz SiK Module to ExLRS Lora based module.
As suggested using a backpack to transfer telemetry from OpenTx via Backpack But we still need a USB based Telemetry Module to connect to Mission Planner.
How do we intend to do that?
Are we planning a new Hardware?[usb based]
I have designed the ExLRS Tx module with usb port planning to port code?
Let me know if this is correct way forward?

**Reply by iottrends** [reply]
I believe Release 3.0 intend to use ESP32 chips paired with 2- SX1280 modules, this makes it an ideal candidate to run CRSF and Mavlink simultaneously.
All the Telemetry data that is coming from Flight Controller can be converted into both CRSF and Mavlink and then transmitted.
Infact in theory Lora is designed to work within group of devices using Lora-group-id, that is how Lora WAN works... No need for multiple Lora Modules.  It can be done with Existing Single Sx1280 modules too!!!.
So if you have 3 devices [Rx, Tx and usb based Tx module connected to Mission-Planner] all 3 of them can communicate, if they are binded together.
Regarding the latency and slow performance...
We already interlace Telemetry data, most people don't understand: while pilots send the fly-control RC signal from TX module ---> Rx module.
and Telemetry data is vice-versa... Rx---> Tx module. Most of them don't use it anyway. we look into our Fatshark googles to view the Telem data coming from our VTx Not the ExLRS Rx modules.
So it would be better use of 500Hz Telem data to go to AutoPilots.

**Reply by commanderguy3001** [reply]
The problem with sending mavlink data instead of telemetry is that we only send a VERY LIMITED amount of telemetry data, just a couple bytes every second, which is not nearly enough for mavlink.

There are two more reasons that I can think of directly why it would be a bad idea:
- many people DO use telemetry and want to use it, and don't want to need to use mavlink to get any telemetry
- the amount of people using telemetry is way bigger than the amount of people that would be using mavlink

The problem of using 3 devices at the same time is that 3 devices don't solve the original issue. Also, you would need to do some kind of time division multiplexing, which introduces complexity. And increase the total bandwidth of the link, as we currently have a duty cycle of somewhere around 80% from what I know, which does not leave enough unused bandwidth for mavlink (this you need to do either way, and it does make the link slower, and leaves less link budget which results in a shorter range, that's just physics).

Also, expresslrs v3 only started to SUPPORT more than one simultaneous sx1280 chip now, it's not "intended to use [multiple]", but rather a: "it is possible at all now".

About your: "In theory lora is designed to work within a group of devices": yes, that is true, but lora is NOT designed to work in a high speed high duty cycle environment, which is what we are doing here.

**Comment by robustini** [comment]
I would say that it would be absolutely fantastic to have this functionality.
Even if it only worked with the system disarmed, useful for changing settings without connecting cables to the flight controls.

**Comment by iottrends** [comment]
Thanks mate... As of now I am taking a simple approach ... to replace the existing Sik Radio modules with ExLRS Rx & Tx[usb based] modules. No relation with existing ExLRS RC modules, I will call this as ExLRS telemetry modules.  
Once the functionality is developed, then we can think of merging/migrating them together.
However one problem persists... 
Telemetry Data is already encoded in CRSF is transmitted which is consumed by OpenTx using yappu script. But still this Telemetry is Receive only... there is noway one can send back the Mavlink commands back to copter/quad via OpenTx. 
Experts/mods can comment on this..

**Reply by commanderguy3001** [reply]
I don't think I'm able to follow the description of your setup fully, but the definition of telemetry is just that a device (the drone/receiver) sends data back to it's master (the transmitter).
Also, CRSF has a specifically defined structure with which you can send pre-defined telemetry fields, which means that it does not send arbitrary data either.
Also: what do you mean with "usb based elrs rx & tx"?
in general there is no way to use elrs to send any mavlink data between your drone and the transmitter in any direction, neither from the drone to the transmitter, nor the other way round.
also, I don't quite understand that one sentence of yours, "No relation with existing ExLRS RC modules, I will call this as ExLRS telemetry modules."

**Reply by robustini** [reply]
Ok, if we can't write then it becomes marginally useful.

**Reply by commanderguy3001** [reply]
I don't quite understand what you mean with "if we can't write [...]"?

**Reply by robustini** [reply]
> I don't quite understand what you mean with "if we can't write [...]"?

"But still this Telemetry is Receive only... there is noway one can send back the Mavlink commands back to copter/quad via OpenTx".

**Reply by commanderguy3001** [reply]
ah, yep, that's the definition of telemetry: sending data back from the receiver to the transmitter
the "normal" direction (sending data from the transmitter to the receiver) is just the standard control link

**Reply by iottrends** [reply]
> I don't think I'm able to follow the description of your setup fully, but the definition of telemetry is just that a device (the drone/receiver) sends data back to it's master (the transmitter). Also, CRSF has a specifically defined structure with which you can send pre-defined telemetry fields, which means that it does not send arbitrary data either. Also: what do you mean with "usb based elrs rx & tx"? in general there is no way to use elrs to send any mavlink data between your drone and the transmitter in any direction, neither from the drone to the transmitter, nor the other way round. also, I don't quite understand that one sentence of yours, "No relation with existing ExLRS RC modules, I will call this as ExLRS telemetry modules."

I am planning to replace the existing Sik Telemetry modules with ExLRS based modules. So one RX module will connect to Telem uarts on the Flight controller and the Another module which connects to Mission Planner/QGnd control  via usb.
That's it...
Using these modules you can do Autonomous missions. same way the SiK Modules do.

**Reply by iottrends** [reply]
> ah, yep, that's the definition of telemetry: sending data back from the receiver to the transmitter the "normal" direction (sending data from the transmitter to the receiver) is just the standard control link

Agree Telem data is one way from Rx Module to the TX module. However if you are using Autopilots like ArduPilot, then you want to send Mavlink commands for waypoint missions from Mission Planner/Qground control.
So there are 2 RX modules on your quad 1. one is regular ExLRS for RC paired up with TX module.
2. Another one is replacement of your Sik Telemetry module [Rx] connected to telemetry ports in quad paired with Ground Module[usb] connected to Mission Planner/Qgrndcontrol.
3. Hope this clarifies.
4.Now Moving further, wishful thinking would be to have a Single ExLRS Rx Module connected to quad. So this Rx Module will  serve mavlink and explrs RC. This Rx is  paired up with 2 Devices on the ground, one is your regular TX module connected to your OpenTX/EdgeTx . Another would be a usb device connected to Qgrndcontrol/Mission Planner.
All 3 devices paired together.

**Reply by commanderguy3001** [reply]
> 

so basically: use expresslrs hardware with different software that is a transparent bidirectional uart instead of a pure CRSF control link.

The problem with only using a single receiver I have stated a couple times already by now.

One more thing you might wanna take a look at to do the first thing: use one additional elrs transmitter and receiver for a transparent uart: https://github.com/ExpressLRS/ExpressLRS/pull/1585
This is the Airport PR by wez, which does exactly that: "a transparent uart over a pair of elrs transmitter and receiver hardware"

**Comment by Pairan** [comment]
linking THIS as it is related to it:
https://github.com/ExpressLRS/Backpack/issues/46

**Reply by commanderguy3001** [reply]
it isn't related all that close, as it would not need a mavlink connection to the drone or anything like that, the only thing you would _need_ to do is to let the backpack output the GPS coordinates from the standard telemetry stream in some form, so it definitely is it's own discussion.

(But I have to admit the original message in this thread was about having a "subset of mavlink sent over the backpack wifi", which _is_ indeed related to the backpack issue you sent)

**Comment by olliw42** [comment]
hey Folks

maybe I may elaborate on the various meanings "MAVLink" can have in the above context. Unfortunately, the word "MAVLink" is used to denote substantially different setups/configurations, and disscusions on "MAVLink" which aren't sufficiently clear on this, as it appears to be often the case, are thus often not very conclusive. I don't use ELRS myself, and really follow it just as a by-stander, but have experience with MAVLink.

Some notation first: The protocol used for the telemetry data over the air is irrelevant for the following. The protocol with which the data are feed into the rx, and with which they are output by the tx, shall hencforth be denoted as "telemetry protocol". 

I think one can (should) distinguish (at least) these three scenarios:

**1. uni-directional link, uni-directional on tx- and rx-sides**

In this scenario a radio link is used with uni-directional telemetry from the rx to the tx. The telemetry protocol can be of any type, such as Frsky or Crsf or passthrough or whatever. The specific telemetry protocol is not relevent to the discussion, as they in principle can always be converted from one to the other by suitable converters. The converter can be an extra unit, like an Arduino, Teensy, Pico, or part of the radio link or radio firmware; there are plenty of such projects out there.

The data provided on the rx-side can also be a MAVLink stream, from which a converter extracts the data in the format suitable for the particular rx. Likewise, a tx may take the received telemetry data and convert it into a MAVLink stream and provide it e.g. on a UART or a BT or whatever.

The crucial point in this setup is that (i) a link with uni-directional telemetry is used, and (ii) that on either or both the tx-side and rx-side a uni-directional MAVLink stream is involved.

This configuration is so to say the simplest "MAVLink" system, and may be sufficient for OSDs, antenna trackers, telemetry apps, and so on. Notably, it is NOT possible to connect with a ground station like MissionPlanner or QGroundControl (vide infra). Also, usually, the MAVLink stream which the rx on the rx-side receives has to be configured "by hand" in the flight controller, but is not set up automatically. The tx may also provide the telemetry data in its native format, such that the telemetry display in the radio or the Yappu telemetry apps also work. That is, this simplest configuration might be perfectly sufficient for what a user may want.

The data bandwidth requirement for the telemetry is low; in fact it is what the radio systems just provide.

This configuration should be easy to realize with ELRS. The converter on the tx side to a MAVLink stream is not so difficult to implement, and should not be so difficult to be integrated into the ELRS tx code.

**2. uni-directional link, bi-directional on tx-side, uni- or bi-directional on rx-side**

In this scenario again a radio link is used with uni-directional telemetry from the rx to the tx. Again, the telemetry protocol can be of any type, such as Frsky or Crsf or passthrough or whatever, doesn't matter here. 

The important difference is on the tx-side, namely that now the tx-side provides a full bi-directional MAVLink connection. The data provided on the rx side can again be of any uni-directional sort, like in case 1, or can in fact also be a full bi-directional MAVLink connection to the flight controller. 

The crucial point in this setup is that while a link with uni-directional telemetry is used, the tx-side provides a full bi-directional MAVLink connection. This (and only this!) makes it possible to connect e.g. with MissionPlanner or QGroundControl, via a uart or BT or wifi. Notably, since the link over the air is only uni-directional, many MAVLink features, such as writing parameters or downloading a mission or controlling a gimbal or camera, and so on, will NOT work in this setup. The tx may agaion also provide the telemetry data in its native format, such that the telemetry display in the radio or the Yappu telemetry apps also work. 

A well-known project which achieves exactly this configuration is Mav2PT.

The data bandwidth requirement for the telemetry is low; in fact it is what the radio systems just provide.

Also this configuration should be - in principle - easy to realize with ELRS. It would imply however that all the code for the full bi-directional MAVLink converter needs to be implemented on the tx-side - one needs so to say integrate Mav2PT into ELRS. In ELRS lingo this would be a backpack feature, I guess.

In my estimante this, or case 1, is what the large majority of those saying "we want MAVLink" actually do want, or are talking about. 

**3. bi-directional link, bi-directional on tx-side, bi-directional on rx-side**

In this scenario a fully bi-directional data channel is realized end-to-end. That is, a radio link is used with bi-directional "telemetry", and both the tx and rx sides offer full bi-directional MAVLink connectivity. 

This bi-directional end-to-end data channel can be "transparent", i.e., work as if it would be just a normal serial/UART connection, or may do some more fancy things to possibly optimize the performance, but the crucial thing is that there are two communication paths, one upstream and one downstream, so that MAVLink packets can flow in both directions. This (and only this!) allows full MAVLink features, including connecting to MissionPlanner/QGroundControl, connecting to arbitrary antenna tracker, writing parameters, downloading missions, and so on.

This is what is widely done in the hobby world with SiK radios, in which case one needs an additional traditional radio link for the rc data, or more (semi)professionally with things like DragonLink, which provides both the radio link and the bi-directional MAVLink data. Also Crossfire allows doing the same. It is also what projects like OpenHD or Ruby offer.  A recent open source project of this kind which is also based on the Semtech (Lora) chips would be mLRS.

The data bandwidth requirements are substantially higher than in case 1 or 2, i.e., significantly higher than what the traditional radio links can provide. In contrast to above voices it is however not that dramatic that Lora, even at 900 MHz, could not work absolutely satisfying (if designed/programmed carefully)(counter example: mLRS).

This configuration - in my estimate too - would not be easy to realize in ELRS, as it indeed would require substantial redesign and rewritting in many places. (this, and the fact that ELRS devs repeatedly expressed that MAVLink is not high on the list, was in fact the primary motivation for mLRS)

-------------------------------------

With this distinction at hand, I think - hope - that it should not be difficult to figure out which of the posters in the above was talking about which scenario. I also think that in any conversion one should be clearer on which scenario it is. 

And I think that especially the ELRS devs should be much much clearer - they IMHO should be crystal clear - which scenario they have in mind in their statements.

In order to maybe help a bit with the conversation, I could suggest this notation

- "**_MAVLink telemetry_**": this are all cases which are build around a uni-directional telemetry link, e.g. scenarios 1 and 2 in the above

- "**_full MAVLink_**": this are the cases which provide full MAVLink operation, i.e., must be based on a bi-directional end-to-end connection with substantial data bandwidth

have fun :)

**Reply by iottrends** [reply]
Well I am planning to experiment... the Mavlink. At first I am not going to use ExLRS code at all, 
I am going to use the ExLRS HW and write basic  Mavlink and see if it works...

**Comment by maxyB19** [comment]
Siyi already has this feature with their FM30 transmitter module. I bought it and it is awesome. Looks like it already exists on the market

**Reply by CorruptName** [reply]
Is this compatible with ELRS Receivers? Or you have to use theirs?

**Reply by radiomistress** [reply]
You have to use theirs. I'm working on using the backpack and a esp15f with the airport branch to achieve the same functionality.

**Reply by ShikOfTheRa** [reply]
@mr-sour I'm interested in using backpack and streaming gps data and altitude. Ideally mavlink but open to anything else really to get that data for antenna tracking. Had  a look at backpack code and it looks easy enough to do that, but have no idea how to get data from ELR into backpack. Be interested in following your dev . 
Edit - I'm only interste din streaming out - not full bi directional

**Comment by NestWolfsteiner** [comment]
"Full MAVLink" will be a game changer for many applications requiring or benefitting from autonomous flight options, often via Ardupilot. 
Effective and efficient build and operating of e.g. camera drones or mapping wings will become available for a much wider community, as the cost of a quality MAVLink connection sometimes add 30% to 70% to the cost of the build.

**Reply by brandonrc** [reply]
You are telling me.. I run RFD900x's on anything I am serious about and those are pricey.

**Comment by nick9one1** [comment]
I agree, MAVlLINK would be great on ELRS. 

ELRS gos hand in hand with long range flying, and long range flying hugely benefits from an automatic antenna tracker. 

Quite a lot of AAT's use the video feed to relay the GPS position, but they only work through analog feeds, not HD (HDZERO, WalkSnail, DJI etc.). 

If ELRS supported MAVLINK, all the above HD platforms would work with the majority of AAT's.

**Comment by olliw42** [comment]
it seems you folks are not aware of mLRS, LoRa optimized for full MAVLink: https://github.com/olliw42/mLRS
:)

**Comment by padcom** [comment]
@olliw42 I think the decision to go STM32 instead of ESP32 was the demise of that project, which is really sad, because it'd be my go-to link for arducopters and it would justify buying second TX to have the 250mW one that I already have and a second one, maybe the big guy from Radiomaster :) But seriously, just port it to ELRS hardware and you'll get massive adoption. That's exactly the kind of project that ELRS is missing.

**Comment by padcom** [comment]
@olliw42 In point of fact ExpressLRS didn't just give us great software but above all the engagement of commercial hardware designers which means there are modules to run the software on. So it became the de-facto hardware platform for opensource control links.

**Comment by olliw42** [comment]
@padcom 
I'm sure aware of what you're saying, but it is not as true as it first seems:
- ELRS receivers can lack transmit power, if you filter them out than there are still some, but only few options left, and none if you want "symmetric" power
- all ELRS receivers lack in pins (this can be somewhat overcome by some workaround, but it's then exactly this, a not so nice workaround)
- all ELRS tx modules today only allow very basic features from the viewpoint of mLRS, i.e., are substantially constraint for the purpose
- you can use the R9 system, the receivers are ok for the purpose, and on the tx side the limitations imposed by the hardware of the tx module are not worse compared to a ELRS tx module, and you can buy it today

so, to the best of my knowledge, as of today, there is no ELRS system which would not imply having to swallow significant compromises

just talking about purhcable ELRS items of course (diy allows to overcome all issues, but then you also can diy mLRS)

note that for full MAVLink you don't want a control link, be it opensource, defacto-standard, or whatever, it's rather the reverse to a control link in terms of direction of largest data flow. This has consequences for the receiver (no low-power) and the protocol.

sorry for missusing this thread here :)

**Comment by padcom** [comment]
@olliw42 You've made some very good points, thank you! Indeed, having a symmetric serial link in MAVLink applications is very much preferable. I really hope your project will catch the attention of hardware designers. I know I'd buy it for my long-range flights. 

That being said, an asymmetric link, with very slow telemetry downlink is still something that can be used in a lot of applications. Not everything needs to fly 20k out and back on full telemetry. Current links are very much OK for ranges up to 5k with full telemetry. Even the very slow baudrate is better than nothing and also better than having a dedicated 2-way telemetry link, like the ones commonly used in Arducopter builds. Having a transparent, but slow and short-range MAVLink with ELRS would still make a lot of people happy and IMHO you could at least try to squeeze your codebase into the ELRS hardware. After all, how far the telemetry will reach is up to the hardware, not software (hardware-configuration aside of course).

**Comment by nick9one1** [comment]
just came across this;

ELRS Airport 

What is Airport?
Airport provides a new firmware option that allows you to turn a regular ExpressLRS TX and RX pair into a bi-directional transparent serial link, over the air.
The TX module is intended to connect via USB to a laptop, and the RX connects to a free UART on your FC, as per usual. This then allows serial comms between the peers, in any protocol you wish to use, for example: MAVLINK (Ardupilot), MSP (Betaflight and INAV), or any other telemetry protocol you might use.

https://github.com/ExpressLRS/ExpressLRS/pull/1904

**Reply by padcom** [reply]
That's what I've had in mind! I love those guys, they make stuff faster than I can come up with ideas 👯

**Comment by FelixSchreiber** [comment]
Can we use the backpack to avoid the USB cable connection. With my TX module I would need to open the case to access the USB port. The backpack could provide the serial link between PC and ELRS Airport TX via WiFi ([ESP-Link](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwjLtq2Ymt7_AhVTlGoFHbdXBrYQFnoECBMQAQ&url=https%3A%2F%2Fgithub.com%2Fjeelabs%2Fesp-link&usg=AOvVaw3s3_slme_wqiaQxZ9WoJms&opi=89978449)?) - although Bluetooth would even be better

**Comment by cbf123** [comment]
For anyone else stumbling over this, you might also want to look at the "mavlink-rc" branch of ELRS.  Basically the flight controller just sees a MAVLink connection, but the RC control signals are sent over the air as usual (low latency) and converted to MAVLink by the receiver.   That way you don't need separate radios for telemetry but you still get low-latency RC controls.

**Reply by ichnograph** [reply]
Does this require a stripped-down TX backpack in the model? Or does it work with a regular receiver as well? Would a pair of 915Mhz nano modules be a good start to test this branch?

**Comment by YrjoB** [comment]
Hello everyone.
I have a question regarding ExpressLRS Airport and Mavlink. When I read that the Airport could connect to the Mission Planner with Mavlink, I decided to try replacing the SIK radio with the ELRS. Radiomaster Bandit micro 900 MHz TX connected to laptop with Missionplanner and in the air a Radiomaster Bandit RP3 RX connected to Pixhawk 6C. Baudrate is 4800 and sometimes everything works OK. Yesterday I flew a few kilometers without losing connection but today the Mavlink does not connect to the aircraft. Do any of you have experience with Airport and Mavlink?.
